packages/component/src/providers/CustomElements/customElements/CodeBlock.ts

```diff
@@ -0,0 +1,240 @@
+      const highlightCodeFragment = this.constructHighlightedCode(code, language, options);
+      this.highlightedCodeFragment = highlightCodeFragment;
+      const body = highlightCodeFragment.querySelector('pre');
+      body?.classList.add('webchat__code-block__body');
+      options.theme && body?.classList.add(options.theme);
+    }
+
+    if (!this.#originalFragment) {
+      this.#originalFragment = document.createDocumentFragment();
+      this.#originalFragment.replaceChildren(...this.children);
+    }
+
+    const sourceFragment = this.highlightedCodeFragment ? this.highlightedCodeFragment : this.#originalFragment;
+    const fragment = sourceFragment?.cloneNode(true);
+
+    this.copyButtonElement ??= this.constructCopyButton();
+    this.copyButtonElement && fragment.insertBefore(this.copyButtonElement, fragment.firstChild);
+
+    this.replaceChildren(fragment);
+  }
+
+  constructCopyButton() {
+    return null;
+  }
+
+  constructHighlightedCode(...args: Parameters<HighlightCodeFn>) {
+    const result = this.highlight(...args);
+    return result instanceof DocumentFragment ? result : parseDocumentFragmentFromString(result);
+  }
+}
+
+const createReactCodeBlockClass = ({
+  codeBlockRef,
+  copyButtonRef,
+  trackCodeBlockRefChanges,
+  trackCopyButtonRefChanges
+}: {
+  codeBlockRef: RefObject<{
+    className: string;
+    highlightCode: HighlightCodeFn;
+    theme: StyleOptions['codeBlockTheme'];
+  }>;
+  copyButtonRef: RefObject<{
+    altCopied: string;
+    altCopy: string;
+    className: string;
+    tagName: string;
+  }>;
+  trackCodeBlockRefChanges: OnTrackFn;
+  trackCopyButtonRefChanges: OnTrackFn;
+}) =>
+  class ReactCodeBlock extends CodeBlock {
+    static observedAttributes = CodeBlock.observedAttributes;
+
+    get options() {
+      // Prefer options from element attributes over ones passed from the context
+      return {
+        theme: codeBlockRef.current.theme,
+        ...super.options
+      };
+    }
+
+    #controller: AbortController;
+    #prevClassName: string;
+
+    connectedCallback() {
+      super.connectedCallback();
+
+      if (!this.#controller) {
+        this.#controller = new AbortController();
+
+        trackCodeBlockRefChanges(() => {
+          this.highlightedCodeFragment = null;
+          this.update();
+        }, this.#controller.signal);
+
+        trackCopyButtonRefChanges(() => {
+          this.copyButtonElement = null;
+          this.update();
+        }, this.#controller.signal);
+      }
+    }
+
+    disconnectedCallback() {
+      this.#controller?.abort();
+      this.#controller = null;
+    }
+
+    highlight(...args: Parameters<HighlightCodeFn>) {
+      const [, language] = args;
+      if (!language) {
+        return super.highlight(...args);
+      }
+      return codeBlockRef.current.highlightCode(...args);
+    }
+
+    updateClassName(): void {
+      this.#prevClassName && this.classList.remove(this.#prevClassName);
```

Comment:
Isn't this simpler if we are using `attributeChangeCallback`?

Reply:
This would mean we need to pass the className to HTML transformer stack and I don't want to do so, as it's kinda dynamic value. For CSS modules we could directly export the class name as it is generated upon package gets built, so in the future we may just import the name.

Reply:
We can also assume the class is static and remove it from the update loop (set upon element creation), this will match the previous html transform approach.

To my mind, if something looks like dynamic, better to include it into the update loop.